README.txt
                    =====================================
                      Package "Bit::Vector" Version 7.3
                    =====================================


This package is available for download either from my web site at

                  http://www.engelschall.com/u/sb/download/

or from any CPAN (= "Comprehensive Perl Archive Network") mirror server:

               http://www.perl.com/CPAN/authors/id/S/ST/STBEY/


Abstract:
---------

Bit::Vector is an efficient C library which allows you to handle
bit vectors, sets (of integers), "big integer arithmetic" and
boolean matrices, all of arbitrary sizes.

The library is efficient (in terms of algorithmical complexity)
and therefore fast (in terms of execution speed) for instance
through the widespread use of divide-and-conquer algorithms.

The package also includes an object-oriented Perl module for
accessing the C library from Perl, and optionally features
overloaded operators for maximum ease of use.

The C library can nevertheless be used stand-alone, without Perl.


What's new in version 7.3:
--------------------------

 +  Fixed t/02_____destroy.t to make it pass on perl-5.17.11


What's new since version 6.5:
-----------------------------

Object persistence:

Since version 6.5, "Bit::Vector" objects can be serialized
and de-serialized automatically with "Storable", out-of-the-box,
without requiring any further user action for this to work.

This is also true for nested data structures (since version 6.8).

See the "Storable" documentation for more details.


Legal issues:
-------------

This package with all its parts is

Copyright (c) 1995 - 2013 by Steffen Beyer.
All rights reserved.

This package is free software; you can use, modify and redistribute
it under the same terms as Perl itself, i.e., under the terms of
the "Artistic License" or the "GNU General Public License".

The C library at the core of this Perl module can additionally
be used, modified and redistributed under the terms of the
"GNU Library General Public License".

Please refer to the files "Artistic.txt", "GNU_GPL.txt" and
"GNU_LGPL.txt" in this distribution, respectively, for details!


Prerequisites:
--------------

Perl version 5.000 or higher, and an ANSI C compiler. (!)
                                     ^^^^^^
Module "Carp::Clan" version 5.3 or higher.

Optionally, module "Storable" version 2.21 or newer.

Note that in order to compile Perl modules which contain
C (and/or XS) code (such as this one), you always HAVE
to use the very same compiler your Perl itself was compiled
with.

Many vendors nowadays ship their operating system already
comprising a precompiled version of Perl. Many times the
compilers used to compile this version of Perl are not
available to or not usually used by the users of these
operating systems.

In such cases building this module (or any other Perl
module containing C and/or XS code) will not work. You
will either have to get the compiler which was used to
compile Perl itself (see for example the section "Compiler:"
in the output of the command "perl -V"), or to build
your own Perl with the compiler of your choice (which
also allows you to take advantage of the various compile-
time switches Perl offers).

Note that Sun Solaris and Red Hat Linux frequently were
reported to suffer from this kind of problem.

Moreover, you usually cannot build any modules under
Windows 95/98 since the Win 95/98 command shell doesn't
support the "&&" operator. You will need the Windows NT
command shell ("cmd.exe") or the "4DOS" shell to be
installed on your Windows 95/98 system first. Note that
Windows NT, Windows 2000 and Windows XP are not affected
and just work fine. I don't know about Windows Vista and
Windows 7, however.

Note that ActiveState provides precompiled binaries of
this module for their Win32 port of Perl ("ActivePerl")
on their web site, which you should be able to install
simply by typing "ppm install Bit-Vector" in your MS-DOS
command shell (but note the "-" instead of "::" in the
package name!). This also works under Windows 95/98 (!).

If your firewall prevents "ppm" from downloading
this package, you can also download it manually from
http://www.activestate.com/ppmpackages/5.005/zips/ or
http://www.activestate.com/ppmpackages/5.6/zips/.
Follow the installation instructions included in
the "zip" archive.


Note to CPAN Testers:
---------------------

After completion, version 7.3 of this module has already
been tested successfully with the following configurations:

  Perl 5.12.3    -  FreeBSD 8.2-PRERELEASE


Installation:
-------------

Please see the file "INSTALL.txt" in this distribution for instructions
on how to install this package.

It is essential that you read this file since one of the special cases
described in it might apply to you, especially if you are running Perl
under Windows.


Changes over previous versions:
-------------------------------

Please refer to the file "CHANGES.txt" in this distribution for a more
detailed version history log.


Documentation:
--------------

The documentation of this package is included in POD format (= "Plain
Old Documentation") in the files with the extension ".pod" in this
distribution, the human-readable markup-language standard for Perl
documentation.

By building this package, this documentation will automatically be
converted into man pages, which will automatically be installed in
your Perl tree for further reference through the installation process,
where they can be accessed by the commands "man Bit::Vector" (Unix)
and "perldoc Bit::Vector" (Unix and Win32 alike), for example.

Available man pages:

    Bit::Vector(3)
    Bit::Vector::Overload(3)
    Bit::Vector::String(3)

If Perl is not available on your system, you can also read the ".pod"
files

    ./Vector.pod
    ./lib/Bit/Vector/Overload.pod
    ./lib/Bit/Vector/String.pod

directly.


What does it do:
----------------

This package implements bit vectors of arbitrary size and provides efficient
methods for handling them.

This goes far beyond the built-in capabilities of Perl for handling bit
vectors (compare with the method list below!).

Moreover, the C core of this package can be used "stand-alone" in other
C applications; Perl is not necessarily required.

The main module of this package is intended to serve as a base class for
other applications or application classes, such as implementing sets or
performing big integer arithmetic.

All methods are implemented in C internally for maximum performance.

An add-on module (named "Bit::Vector::Overload") provides overloaded
arithmetic and relational operators for maximum ease of use (Perl only).

Note that there is (of course) a little speed penalty to pay for
overloaded operators. If speed is crucial, use the "Bit::Vector"
module alone!

Another module, "Bit::Vector::String", offers additional methods
for generic string export and import (also Perl only).

This package is useful for a large range of different tasks:

  -  For example for implementing sets and performing set operations
     (like union, difference, intersection, complement, check for subset
     relationship etc.),

  -  as a basis for many efficient algorithms, for instance the
     "Sieve of Erathostenes" (for calculating prime numbers),

     (The complexities of the methods in this package are usually either
      O(1) or O(n/b), where "b" is the number of bits in a machine word
      on your system.)

  -  for shift registers of arbitrary length (for example for cyclic
     redundancy checksums),

  -  to calculate "look-ahead", "first" and "follow" character sets
     for parsers and compiler-compilers,

  -  for graph algorithms,

  -  for efficient storage and retrieval of status information,

  -  for performing text synthesis ruled by boolean expressions,

  -  for "big integer" arithmetic with arbitrarily large integers,

  -  for manipulations of chunks of bits of arbitrary size,

  -  for bitwise processing of audio CD wave files,

  -  to convert formats of data files,

and more.

(A number of example applications is available from my web site at
http://www.engelschall.com/u/sb/download/.)

A large number of import/export methods allow you to access individual
bits, contiguous ranges of bits, machine words, arbitrary chunks of
bits, lists (arrays) of chunks of bits or machine words and a whole
bit vector at once (for instance for blockwrite/-read to and from
a file).

You can also import and export the contents of a bit vector in binary,
hexadecimal and decimal representation as well as ".newsrc" style
enumerations.

Note that this package is specifically designed for efficiency, which is
also the reason why its methods are implemented in C.

To further increase execution speed, the package doesn't use bytes as its
basic storage unit, but rather uses machine words, assuming that a machine
word is the most efficiently handled size of all scalar types on all
machines (that's what the ANSI C standard proposes and assumes anyway).

In order to achieve this, it automatically determines the number of bits
in a machine word on your system and then adjusts its internal configuration
constants accordingly.

The greater the size of this basic storage unit, the better the complexity
(= execution speed) of the methods in this package, but also the greater the
average waste of unused bits in the last word.

The range of available methods is exceptionally large for this kind of library;
in detail:


CHANGES.txt
                    =====================================
                      Package "Bit::Vector" Version 7.3
                    =====================================


                 Copyright (c) 1995 - 2013 by Steffen Beyer.
                             All rights reserved.


Version history:
----------------

Version 7.3   01.06.2013

 +  Fixed t/02_____destroy.t to make it pass on perl-5.17.11

Version 7.2   17.05.2012

 +  Changed the definition of the "boolean" data type in ToolBox.h

Version 7.1   29.09.2009

 +  Added prefix "BV_" to all global identifiers in "BitVector.c",
    "BitVector.h" and "Vector.xs"

Version 7.0   22.08.2009

 +  Fixed a bug in "GCD2()" in "BitVector.c" with null-size vectors
 +  Added more test cases for "GCD()" in "t/17_________gcd.t"
 +  Almost completely rewrote "BitVector_Boot()" in "BitVector.c"
    in order to fix the problem that it calculated an unsigned
    long to have 33 bits under MacOS X (Darwin), leading to
    failing tests in "t/28___chunklist.t"
 +  BEWARE that the type "ErrCode", which is part of the inter-
    face of the C library "BitVector.c", has been changed!
 +  Added more tests and a debugging aid to "t/28___chunklist.t"
 +  Removed the two example files showing how to freeze/thaw
    "Bit::Vector" objects using "Data::Dumper", because after
    closer inspection this proved to be a can of worms

Version 6.9   12.08.2009

 +  Replaced "Storable::freeze()" by "Storable::nfreeze()"
 +  Added more test cases for "Storable"
 +  Added two example files showing how to freeze/thaw
    using "Data::Dumper"

Version 6.8   10.08.2009

 +  Fixed the bug of "Storable" not reconstructing nested data
    structures properly which contain several references to
    the same "Bit::Vector" object
 +  Fixed hard-coded table size "MASKTABSIZE" in "BitVector.c"
 +  Small cosmetic fixes to the documentation

Version 6.7   08.08.2009

 +  Replaced "STORABLE_thaw()" by "STORABLE_attach()"

Version 6.6   27.07.2009

 +  Made the module thread-safe and MacOS X compatible
 +  Prevented "BitVector_Boot()" in "BitVector.c" to leak
    memory when called repeatedly

Version 6.5   27.07.2009

 +  Added automatic support for module "Storable"

Version 6.4   03.10.2004

 +  Added compiler directives for C++.
 +  Improved the method "Norm()".
 +  Removed "Carp::Clan" from the distribution (available separately).
 +  Added "Bit::Vector::String" for generic string import/export functions.
 +  Added a new test file "t/40__auxiliary.t" for "Bit::Vector::String".
 +  Fixed a bug in method "Copy()" concerning sign (MSB) extension.

Version 6.3   28.09.2002

 +  Added "Create_List()" and "GCD2()" in "BitVector.c".
 +  "new()" now can optionally return a list of bit vectors.
 +  "GCD()" now can optionally return the two integer factors
    "x" and "y" for the linear combination of its input values
    "a" and "b" so that gcd(a,b) = x * a + y * b.
 +  Changed the test files "t/01________new.t" and "t/09_parameters.t"
    as well as the documentation accordingly.
 +  Added a new test file "t/17________gcd.t".
 +  Further simplified the error handlers in "Vector.xs", making the
    resulting object library file substantially smaller (about 20%!)
    and thus faster to load.

Version 6.2   15.09.2002

 +  Fixed missing "\" in regular expression in "t/00____version.t".
 +  Overhauled the error handlers in "BitVector.c" and "Vector.xs".
 +  Added a demo "examples/test.c" to play around with the C library.
 +  Attempted to fix all compiler warnings.
 +  Fixed a problem in test "t/01________new.t" with Perl using
    64 bit integers; now contains "use integer;" to avoid failure.
 +  Synchronized "Carp::Clan" and "ToolBox.h" with "Date::Calc" 5.2.
 +  Revealed common method aliases (Or, And, etc.) in the documentation.
 +  Fixed GCD(), now GCD(z,0) == GCD(0,z) == z and negative values
    are handled correctly.

Version 6.1   08.10.2001

 +  Fixed the problem which caused some of the tests with overloaded
    operators to fail under Perl 5.7.1 and Perl 5.7.2.
 +  Added the module "Carp::Clan" to the distribution.
 +  Changed most of the error messages in Bit::Vector::Overload.
 +  Changed the test script "t/30_overloaded.t" accordingly.
 +  Changed "Makefile.PL" to write a "patchlevel.h" file so that
    "Vector.xs" can automatically detect the Perl version and
    whether to use the macro PL_na or na, GIMME_V or GIMME.
    Tweaking the code before installing is not necessary anymore.
 +  Changed the documentation in files README.txt and INSTALL.txt
    accordingly.
 +  Changed the internal names BitVector_msb and BitVector_lsb to
    BitVector_msb_ and BitVector_lsb_ in order to make the (non-

BitVector.h
#ifndef MODULE_BIT_VECTOR
#define MODULE_BIT_VECTOR
#ifdef __cplusplus
extern "C"
{
#endif
/*****************************************************************************/
/*  MODULE NAME:  BitVector.h                           MODULE TYPE:  (adt)  */
/*****************************************************************************/
/*  MODULE IMPORTS:                                                          */
/*****************************************************************************/
#include <stdlib.h>                                 /*  MODULE TYPE:  (sys)  */
#include <limits.h>                                 /*  MODULE TYPE:  (sys)  */
#include <string.h>                                 /*  MODULE TYPE:  (sys)  */
#include <ctype.h>                                  /*  MODULE TYPE:  (sys)  */
#include "ToolBox.h"                                /*  MODULE TYPE:  (dat)  */
/*****************************************************************************/
/*  MODULE INTERFACE:                                                        */
/*****************************************************************************/

typedef enum
{
    BV_ErrCode_Ok = 0,     /* everything went allright                       */

    BV_ErrCode_Type,       /* types word and size_t have incompatible sizes  */
    BV_ErrCode_Bits,       /* bits of word and sizeof(word) are inconsistent */
    BV_ErrCode_Word,       /* size of word is less than 16 bits              */
    BV_ErrCode_Powr,       /* number of bits of word is not a power of two   */
    BV_ErrCode_Loga,       /* error in calculation of logarithm              */
    BV_ErrCode_Lpwr,       /* number of bits of long is not a power of two   */
    BV_ErrCode_WgtL,       /* size of word is greater than size of long      */

    BV_ErrCode_Null,       /* unable to allocate memory                      */

    BV_ErrCode_Indx,       /* index out of range                             */
    BV_ErrCode_Ordr,       /* minimum > maximum index                        */
    BV_ErrCode_Size,       /* bit vector size mismatch                       */
    BV_ErrCode_Pars,       /* input string syntax error                      */
    BV_ErrCode_Ovfl,       /* numeric overflow error                         */
    BV_ErrCode_Same,       /* operands must be distinct                      */
    BV_ErrCode_Expo,       /* exponent must be positive                      */
    BV_ErrCode_Zero,       /* division by zero error                         */
    BV_ErrCode_Oops        /* unexpected error (contact author)              */
}
           BV_ErrCode;

typedef
wordptr   *bv_listptr;

/* ===> MISCELLANEOUS BASIC FUNCTIONS: <=== */

charptr    BitVector_Error      (BV_ErrCode error); /* map errcode to string */

BV_ErrCode BitVector_Boot       (void);             /* 0 = ok, 1..17 = error */

N_word     BitVector_Size       (N_int bits); /* bit vec. size (# of words)  */
N_word     BitVector_Mask       (N_int bits); /* bit vec. mask (unused bits) */

/* ===> CLASS METHODS: <=== */

charptr    BitVector_Version    (void);             /* return version string */

N_int      BitVector_Word_Bits  (void);  /* return # of bits in machine word */
N_int      BitVector_Long_Bits  (void); /* return # of bits in unsigned long */

/* ===> CONSTRUCTOR METHODS: <=== */

wordptr    BitVector_Create     (N_int bits, boolean clear);       /* malloc */
bv_listptr BitVector_Create_List(N_int bits, boolean clear, N_int count);

wordptr    BitVector_Resize     (wordptr oldaddr, N_int bits);    /* realloc */

wordptr    BitVector_Shadow     (wordptr addr);  /* make same size but empty */
wordptr    BitVector_Clone      (wordptr addr);      /* make exact duplicate */

wordptr    BitVector_Concat     (wordptr X, wordptr Y);   /* return concat'd */

/* ===> DESTRUCTOR METHODS: <=== */

void       BitVector_Dispose            (charptr string);          /* string */
void       BitVector_Destroy            (wordptr addr);            /* bitvec */
void       BitVector_Destroy_List       (bv_listptr list, N_int count); /* list */

/* ===> OBJECT METHODS: <=== */

/* ===> bit vector copy function: */

void       BitVector_Copy       (wordptr X, wordptr Y);           /* X = Y   */

/* ===> bit vector initialization: */

void       BitVector_Empty      (wordptr addr);                   /* X = {}  */
void       BitVector_Fill       (wordptr addr);                   /* X = ~{} */
void       BitVector_Flip       (wordptr addr);                   /* X = ~X  */

void       BitVector_Primes     (wordptr addr);

/* ===> miscellaneous functions: */

void       BitVector_Reverse    (wordptr X, wordptr Y);

/* ===> bit vector interval operations and functions: */

void       BitVector_Interval_Empty     (wordptr addr, N_int lower, N_int upper);
void       BitVector_Interval_Fill      (wordptr addr, N_int lower, N_int upper);
void       BitVector_Interval_Flip      (wordptr addr, N_int lower, N_int upper);
void       BitVector_Interval_Reverse   (wordptr addr, N_int lower, N_int upper);

boolean    BitVector_interval_scan_inc  (wordptr addr, N_int start,
                                         N_intptr min, N_intptr max);
boolean    BitVector_interval_scan_dec  (wordptr addr, N_int start,
                                         N_intptr min, N_intptr max);

void       BitVector_Interval_Copy      (wordptr X, wordptr Y, N_int Xoffset,
                                         N_int Yoffset, N_int length);

wordptr    BitVector_Interval_Substitute(wordptr X, wordptr Y,
                                         N_int Xoffset, N_int Xlength,
                                         N_int Yoffset, N_int Ylength);

/* ===> bit vector test functions: */

boolean    BitVector_is_empty   (wordptr addr);               /* X == {} ?   */
boolean    BitVector_is_full    (wordptr addr);               /* X == ~{} ?  */

boolean    BitVector_equal      (wordptr X, wordptr Y);       /* X == Y ?    */
Z_int      BitVector_Lexicompare(wordptr X, wordptr Y);       /* X <,=,> Y ? */
Z_int      BitVector_Compare    (wordptr X, wordptr Y);       /* X <,=,> Y ? */

/* ===> bit vector string conversion functions: */

charptr    BitVector_to_Hex     (wordptr addr);
BV_ErrCode BitVector_from_Hex   (wordptr addr, charptr string);

charptr    BitVector_to_Bin     (wordptr addr);
BV_ErrCode BitVector_from_Bin   (wordptr addr, charptr string);

charptr    BitVector_to_Dec     (wordptr addr);
BV_ErrCode BitVector_from_Dec   (wordptr addr, charptr string);

charptr    BitVector_to_Enum    (wordptr addr);
BV_ErrCode BitVector_from_Enum  (wordptr addr, charptr string);

/* ===> bit vector bit operations, functions & tests: */

void       BitVector_Bit_Off    (wordptr addr, N_int index); /*  X = X \ {x} */
void       BitVector_Bit_On     (wordptr addr, N_int index); /*  X = X + {x} */
boolean    BitVector_bit_flip   (wordptr addr, N_int index); /* (X+{x})\(X*{x}) */

boolean    BitVector_bit_test   (wordptr addr, N_int index); /*  {x} in X ?  */

void       BitVector_Bit_Copy   (wordptr addr, N_int index, boolean bit);

/* ===> bit vector bit shift & rotate functions: */

void       BitVector_LSB                (wordptr addr, boolean bit);
void       BitVector_MSB                (wordptr addr, boolean bit);
boolean    BitVector_lsb_               (wordptr addr);
boolean    BitVector_msb_               (wordptr addr);
boolean    BitVector_rotate_left        (wordptr addr);
boolean    BitVector_rotate_right       (wordptr addr);
boolean    BitVector_shift_left         (wordptr addr, boolean carry_in);
boolean    BitVector_shift_right        (wordptr addr, boolean carry_in);
void       BitVector_Move_Left          (wordptr addr, N_int bits);
void       BitVector_Move_Right         (wordptr addr, N_int bits);

/* ===> bit vector insert/delete bits: */

void       BitVector_Insert     (wordptr addr, N_int offset, N_int count,
                                 boolean clear);
void       BitVector_Delete     (wordptr addr, N_int offset, N_int count,
                                 boolean clear);

/* ===> bit vector arithmetic: */

boolean    BitVector_increment  (wordptr addr);                     /*  X++  */
boolean    BitVector_decrement  (wordptr addr);                     /*  X--  */

boolean    BitVector_compute    (wordptr X, wordptr Y, wordptr Z, boolean minus,
                                                                  boolean *carry);
boolean    BitVector_add        (wordptr X, wordptr Y, wordptr Z, boolean *carry);
boolean    BitVector_sub        (wordptr X, wordptr Y, wordptr Z, boolean *carry);
boolean    BitVector_inc        (wordptr X, wordptr Y);
boolean    BitVector_dec        (wordptr X, wordptr Y);

void       BitVector_Negate     (wordptr X, wordptr Y);
void       BitVector_Absolute   (wordptr X, wordptr Y);
Z_int      BitVector_Sign       (wordptr addr);
BV_ErrCode BitVector_Mul_Pos    (wordptr X, wordptr Y, wordptr Z, boolean strict);
BV_ErrCode BitVector_Multiply   (wordptr X, wordptr Y, wordptr Z);
BV_ErrCode BitVector_Div_Pos    (wordptr Q, wordptr X, wordptr Y, wordptr R);
BV_ErrCode BitVector_Divide     (wordptr Q, wordptr X, wordptr Y, wordptr R);
BV_ErrCode BitVector_GCD        (wordptr X, wordptr Y, wordptr Z);
BV_ErrCode BitVector_GCD2       (wordptr U, wordptr V, wordptr W,   /*   O   */
                                            wordptr X, wordptr Y);  /*   I   */
BV_ErrCode BitVector_Power      (wordptr X, wordptr Y, wordptr Z);

/* ===> direct memory access functions: */

void       BitVector_Block_Store(wordptr addr, charptr buffer, N_int length);
charptr    BitVector_Block_Read (wordptr addr, N_intptr length);

/* ===> word array functions: */

void       BitVector_Word_Store (wordptr addr, N_int offset, N_int value);
N_int      BitVector_Word_Read  (wordptr addr, N_int offset);

void       BitVector_Word_Insert(wordptr addr, N_int offset, N_int count,
                                 boolean clear);
void       BitVector_Word_Delete(wordptr addr, N_int offset, N_int count,
                                 boolean clear);

/* ===> arbitrary size chunk functions: */

void       BitVector_Chunk_Store(wordptr addr, N_int chunksize,
                                 N_int offset, N_long value);
N_long     BitVector_Chunk_Read (wordptr addr, N_int chunksize,
                                 N_int offset);

/* ===> set operations: */

void       Set_Union            (wordptr X, wordptr Y, wordptr Z); /* X = Y + Z */
void       Set_Intersection     (wordptr X, wordptr Y, wordptr Z); /* X = Y * Z */
void       Set_Difference       (wordptr X, wordptr Y, wordptr Z); /* X = Y \ Z */
void       Set_ExclusiveOr      (wordptr X, wordptr Y, wordptr Z); /*(Y+Z)\(Y*Z)*/
void       Set_Complement       (wordptr X, wordptr Y);            /* X = ~Y */

/* ===> set functions: */

boolean    Set_subset           (wordptr X, wordptr Y);            /* X in Y */

N_int      Set_Norm             (wordptr addr);                    /* = |X|  */
N_int      Set_Norm2            (wordptr addr);                    /* = |X|  */
N_int      Set_Norm3            (wordptr addr);                    /* = |X|  */
Z_long     Set_Min              (wordptr addr);                    /* = min(X) */
Z_long     Set_Max              (wordptr addr);                    /* = max(X) */

/* ===> matrix-of-booleans operations: */

void       Matrix_Multiplication(wordptr X, N_int rowsX, N_int colsX,
                                 wordptr Y, N_int rowsY, N_int colsY,
                                 wordptr Z, N_int rowsZ, N_int colsZ);

void       Matrix_Product       (wordptr X, N_int rowsX, N_int colsX,
                                 wordptr Y, N_int rowsY, N_int colsY,
                                 wordptr Z, N_int rowsZ, N_int colsZ);

void       Matrix_Closure       (wordptr addr, N_int rows, N_int cols);

void       Matrix_Transpose     (wordptr X, N_int rowsX, N_int colsX,
                                 wordptr Y, N_int rowsY, N_int colsY);

/*****************************************************************************/
/*  MODULE RESOURCES:                                                        */
/*****************************************************************************/

#define BV_BITS_(BitVector) *(BitVector-3)
#define BV_SIZE_(BitVector) *(BitVector-2)
#define BV_MASK_(BitVector) *(BitVector-1)

#define BV_ERRCODE_TYPE  "sizeof(word) > sizeof(size_t)"
#define BV_ERRCODE_BITS  "bits(word) != sizeof(word)*8"
#define BV_ERRCODE_WORD  "bits(word) < 16"
#define BV_ERRCODE_POWR  "bits(word) is not a power of two"
#define BV_ERRCODE_LOGA  "bits(word) != 2^ld(bits(word))"
#define BV_ERRCODE_LPWR  "bits(long) is not a power of two"
#define BV_ERRCODE_WGTL  "bits(word) > bits(long)"
#define BV_ERRCODE_NULL  "unable to allocate memory"
#define BV_ERRCODE_INDX  "index out of range"
#define BV_ERRCODE_ORDR  "minimum > maximum index"
#define BV_ERRCODE_SIZE  "bit vector size mismatch"
#define BV_ERRCODE_PARS  "input string syntax error"
#define BV_ERRCODE_OVFL  "numeric overflow error"
#define BV_ERRCODE_SAME  "result vector(s) must be distinct"
#define BV_ERRCODE_EXPO  "exponent must be positive"
#define BV_ERRCODE_ZERO  "division by zero error"
#define BV_ERRCODE_OOPS  "unexpected internal error - please contact author"

extern const N_int BV_ByteNorm[256];
/*
{
    0x00, 0x01, 0x01, 0x02,  0x01, 0x02, 0x02, 0x03,
    0x01, 0x02, 0x02, 0x03,  0x02, 0x03, 0x03, 0x04, // 0x00 //
    0x01, 0x02, 0x02, 0x03,  0x02, 0x03, 0x03, 0x04,
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05, // 0x10 //
    0x01, 0x02, 0x02, 0x03,  0x02, 0x03, 0x03, 0x04,
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05, // 0x20 //
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05,
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06, // 0x30 //
    0x01, 0x02, 0x02, 0x03,  0x02, 0x03, 0x03, 0x04,
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05, // 0x40 //
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05,
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06, // 0x50 //
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05,
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06, // 0x60 //
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06,
    0x04, 0x05, 0x05, 0x06,  0x05, 0x06, 0x06, 0x07, // 0x70 //
    0x01, 0x02, 0x02, 0x03,  0x02, 0x03, 0x03, 0x04,
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05, // 0x80 //
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05,
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06, // 0x90 //
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05,
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06, // 0xA0 //
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06,
    0x04, 0x05, 0x05, 0x06,  0x05, 0x06, 0x06, 0x07, // 0xB0 //
    0x02, 0x03, 0x03, 0x04,  0x03, 0x04, 0x04, 0x05,
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06, // 0xC0 //
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06,
    0x04, 0x05, 0x05, 0x06,  0x05, 0x06, 0x06, 0x07, // 0xD0 //
    0x03, 0x04, 0x04, 0x05,  0x04, 0x05, 0x05, 0x06,
    0x04, 0x05, 0x05, 0x06,  0x05, 0x06, 0x06, 0x07, // 0xE0 //
    0x04, 0x05, 0x05, 0x06,  0x05, 0x06, 0x06, 0x07,
    0x05, 0x06, 0x06, 0x07,  0x06, 0x07, 0x07, 0x08  // 0xF0 //
};
*/

/*****************************************************************************/
/*  MODULE IMPLEMENTATION:                                                   */
/*****************************************************************************/

/*****************************************************************************/
/*  VERSION:  7.3                                                            */
/*****************************************************************************/
/*  VERSION HISTORY:                                                         */
/*****************************************************************************/
/*                                                                           */
/*    Version 7.3  01.06.13  No changes.                                     */
/*    Version 7.2  17.05.12  No changes.                                     */
/*    Version 7.1  29.09.09  Added prefix "BV_" to all global identifiers.   */
/*    Version 7.0  22.08.09  Fixed bugs in "GCD2()" and "Boot()".            */
/*    Version 6.9  12.08.09  Removed an obsolete warning (memory leak).      */
/*    Version 6.8  10.08.09  Fixed hard-coded table size BV_MASKTABSIZE.     */
/*    Version 6.7  08.08.09  No changes.                                     */
/*    Version 6.6  27.07.09  Made it thread-safe and MacOS X compatible.     */
/*    Version 6.5  27.07.09  Added automatic support for module "Storable".  */
/*    Version 6.4  03.10.04  Added C++ comp. directives. Improved "Norm()".  */
/*    Version 6.3  28.09.02  Added "Create_List()" and "GCD2()".             */
/*    Version 6.2  15.09.02  Overhauled error handling. Fixed "GCD()".       */
/*    Version 6.1  08.10.01  Make VMS linker happy: _lsb,_msb => _lsb_,_msb_ */
/*    Version 6.0  08.10.00  Corrected overflow handling.                    */
/*    Version 5.8  14.07.00  Added "Power()". Changed "Copy()".              */
/*    Version 5.7  19.05.99  Quickened "Div_Pos()". Added "Product()".       */
/*    Version 5.6  02.11.98  Leading zeros eliminated in "to_Hex()".         */
/*    Version 5.5  21.09.98  Fixed bug of uninitialized "error" in Multiply. */
/*    Version 5.4  07.09.98  Fixed bug of uninitialized "error" in Divide.   */
/*    Version 5.3  12.05.98  Improved Norm. Completed history.               */
/*    Version 5.2  31.03.98  Improved Norm.                                  */
/*    Version 5.1  09.03.98  No changes.                                     */
/*    Version 5.0  01.03.98  Major additions and rewrite.                    */
/*    Version 4.2  16.07.97  Added is_empty, is_full.                        */
/*    Version 4.1  30.06.97  Added word-ins/del, move-left/right, inc/dec.   */
/*    Version 4.0  23.04.97  Rewrite. Added bit shift and bool. matrix ops.  */
/*    Version 3.2  04.02.97  Added interval methods.                         */
/*    Version 3.1  21.01.97  Fixed bug on 64 bit machines.                   */
/*    Version 3.0  12.01.97  Added flip.                                     */
/*    Version 2.0  14.12.96  Efficiency and consistency improvements.        */
/*    Version 1.1  08.01.96  Added Resize and ExclusiveOr.                   */
/*    Version 1.0  14.12.95  First version under UNIX (with Perl module).    */
/*    Version 0.9  01.11.93  First version of C library under MS-DOS.        */
/*    Version 0.1  ??.??.89  First version in Turbo Pascal under CP/M.       */
/*                                                                           */
/*****************************************************************************/
/*  AUTHOR:                                                                  */
/*****************************************************************************/
/*                                                                           */
/*    Steffen Beyer                                                          */
/*    mailto:STBEY@cpan.org                                                  */
/*    http://www.engelschall.com/u/sb/download/                              */
/*                                                                           */
/*****************************************************************************/
/*  COPYRIGHT:                                                               */
/*****************************************************************************/
/*                                                                           */
/*    Copyright (c) 1995 - 2013 by Steffen Beyer.                            */
/*    All rights reserved.                                                   */
/*                                                                           */
/*****************************************************************************/
/*  LICENSE:                                                                 */
/*****************************************************************************/
/*                                                                           */
/*    This library is free software; you can redistribute it and/or          */
/*    modify it under the terms of the GNU Library General Public            */
/*    License as published by the Free Software Foundation; either           */
/*    version 2 of the License, or (at your option) any later version.       */
/*                                                                           */
/*    This library is distributed in the hope that it will be useful,        */
/*    but WITHOUT ANY WARRANTY; without even the implied warranty of         */
/*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       */
/*    Library General Public License for more details.                       */
/*                                                                           */
/*    You should have received a copy of the GNU Library General Public      */
/*    License along with this library; if not, write to the                  */
/*    Free Software Foundation, Inc.,                                        */
/*    59 Temple Place, Suite 330, Boston, MA 02111-1307 USA                  */
/*                                                                           */
/*    or download a copy from ftp://ftp.gnu.org/pub/gnu/COPYING.LIB-2.0      */
/*                                                                           */
/*****************************************************************************/
#ifdef __cplusplus
}
#endif
#endif
